### DISCUSSION TITLE
EIP165 in EIP721

### DISCUSSION THREAD

**Discussion by andrew-fleming** [OpenZeppelin/cairo-contracts#71] [Design]
What is the best path forward regarding the incorporation of [EIP165](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-165.md) with [EIP721](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md) (and subsequently, in general)? Being that StarkNet handles selectors by using the function name (versus the signature with EVM), how should we best approach implementing EIP165? A big thanks to @dewi-tim for bringing this up and offering the following options to consider. 

Should we:

> 1. Attempt to replicate Solidity version behavior exactly (i.e. include types when computing selector, somehow imitate bytes4, use XOR)
> 2. Use StarkNet selectors (felts derived from function name only) and use bitwise xor builtins on felts)
> 3. Use StarkNet selectors and a different operation to compute interfaceID (e.g. multiplication/addition) <- may have different properties to XOR

My initial response leaned towards thinking that using the builtins on felts would be cheaper; however, regular arithmetic to compute the interfaceId may be less costly (if we don't know for certain, could be a good little research project). Or, of course, should we aim towards EVM compatibility?  What other implications should be considered?

An important aspect to contemplate as well is @martriay 's comment in [ERC20 function naming](https://github.com/OpenZeppelin/cairo-contracts/discussions/34#discussioncomment-1655945)—namely, the meta of should these standards in Cairo be interoperable with EVM or should they move away from EVM. 

What do you guys think?

**Comment by frangio** [comment]
One option that I think you didn't mention is that we could just copy the inteface identifier constant verbatim, instead of recomputing it in Cairo. If we're going with implementing ERCs to the letter (see the discussion about camel case), I think this is the only option that is coherent with that approach.

**Reply by andrew-fleming** [reply]
Excellent point!
